Attested TLS openssl samples. (#3750)
openenclave/openenclave: samples/attested_tls/README.md

 The audience is assumed to be familiar:
 [What is an Attested TLS channel](AttestedTLSREADME.md#what-is-an-attested-tls-channel)

# The Attested TLS sample

It has the following properties:

- Demonstrates attested TLS feature
  - between two enclaves
  - between an enclave application and a non enclave application
- Use of mbedTLS/OpenSSL crypto libraries within enclaves for TLS

> - Developer can select mbedTLS/OpenSSL crypto library by setting build time configuration variable `OE_CRYPTO_LIB` in Makefiles [here 1](Makefile#L8) or [here 2](CMakeLists.txt#L10) to `mbedtls` or `openssl`. Note that `OE_CRYPTO_LIB` is case sensitive and should be set to either `mbedtls` or `openssl`.

Reviewer: ```suggestion
- Select between Mbed TLS and OpenSSL crypto library by setting an environment variable  `OE_CRYPTO_LIB` to `mbedtls` or `openssl` at  build time (Note that `OE_CRYPTO_LIB` is case sensitive).
  - If the `OE_CRYPTO_LIB` not set, Mbed TLS will be used by default (see [Makefile](Makefile#L8) and [CMakeLists.txt](CMakeLists.txt#L10)).
```

> - Please check recommended configurations and practices to follow [here](#openssl-recommended-configurations-for-tls-communication) when using OpenSSL crypto library.

Reviewer: ```suggestion
- Configure the server with recommended cipher suites and elliptic curves (refer to the [section](#openssl-recommended-configurations-for-tls-communication) for more details).
```

> - Enclave APIs used:q

Reviewer 2: Is `:q` is a typo?

> - Enclave APIs used:q

Reviewer: ```suggestion
- Use of following Enclave APIs
```

To build and run the samples, refer to documentation in the main [README file](../README.md#building-the-samples.md).

> Note: This sample uses an OE SDK customized version of mbedtls library for TLS channel connection when `OE_CRYPTO_LIB` is set to `mbedtls` in Makefiles. It has MBEDTLS_NET_C component enabled, which has a dependency on the newly added [socket support](../../docs/UsingTheIOSubsystem.md#socketh) in 0.6.0 OE SDK release (for more details see [Using the Open Enclave I/O subsystem](../../docs/UsingTheIOSubsystem.md#opting-in) for details). So in order to build successfully, you would need to link with liboehostsock and libhostresolver libraries to satisfy the dependency.

Reviewer: ```suggestion
Note: This sample uses an OE SDK customized version of Mbed TLS library for TLS channel connection when `OE_CRYPTO_LIB` is set to `mbedtls`. It has MBEDTLS_NET_C component enabled, which depends on the newly added [socket support](../../docs/UsingTheIOSubsystem.md#socketh) in 0.6.0 OE SDK release (refer to [Using the Open Enclave I/O subsystem](../../docs/UsingTheIOSubsystem.md#opting-in) for more details). To build successfully, you would need to link with liboehostsock and libhostresolver libraries to satisfy the dependency.
```

```cmd
.\server\host\tls_server_host .\server\enc\tls_server_enc.signed -port:12341 -server-in-loop
```
### OpenSSL recommended configurations for TLS communication

>   Open Enclave's security guidance recommends to use only subset of cipher suites, elliptic curve algorithms and TLS protocols available in the OpenSSL library. The recommended algorithms and protocols are listed below. Enclave application developers can use [`initalize_ssl_context`](common/openssl_utility.cpp#L118) method to configure the SSL_CTX and limit the algorithms and protocols to below list. In Attested TLS sample application, server and client modules invoke the method `initalize_ssl_context` [here](server/enc/openssl_server.cpp#L147) and [here](client/enc/openssl_client.cpp#L200) respectively to account for the mentioned recommendations. It is strongly recommended that the application developers uses similar method to configure the SSL_CTX and limit the available ciphersuites and elliptic curve algorithms to below list.

Reviewer: ```suggestion
  Open Enclave's security guidance recommends to use only a subset of cipher suites in TLS 1.2/1.3 and elliptic curve algorithms available in the OpenSSL library. The [`initalize_ssl_context`](common/openssl_utility.cpp#L118) demonstrates how to configure both a [server](server/enc/openssl_server.cpp#L147) and a [client](client/enc/openssl_client.cpp#L200) with the recommended algorithms and protocols (see the full list below).
```

Author: I tried it and looks like existing mbedtls application runs with TLS 1.2. With the TLS 1.2 disabled, the communication between enclave mbedtls server and non_enclave openssl client failing with below error. 
`! mbedtls_ssl_handshake returned -0x7380
TLS server: server communication error -29568
TLS server: Last error was: -29568 - SSL - The server has no ciphersuites in common with the client`

Reviewer: I guess you're replying to my question above. I see problem is that non_enc client is also expected to work with Mbed TLS server, correct? In that case, we can just leave as it is.

Author: Yes, I could not reply in there for some reason. Hence added comment here. Yes, we will leave as it is.

>   Open Enclave's security guidance recommends to use only subset of cipher suites, elliptic curve algorithms and TLS protocols available in the OpenSSL library. The recommended algorithms and protocols are listed below. Enclave application developers can use [`initalize_ssl_context`](common/openssl_utility.cpp#L118) method to configure the SSL_CTX and limit the algorithms and protocols to below list. In Attested TLS sample application, server and client modules invoke the method `initalize_ssl_context` [here](server/enc/openssl_server.cpp#L147) and [here](client/enc/openssl_client.cpp#L200) respectively to account for the mentioned recommendations. It is strongly recommended that the application developers uses similar method to configure the SSL_CTX and limit the available ciphersuites and elliptic curve algorithms to below list.

Reviewer 3: ```suggestion
  It is strongly recommended that developers configure OpenSSL to restrict the set TLS versions, ciphersuites and elliptic curves to be used for TLS connections to enclave:
  
    <configuration list from lines 99-117> 

This sample illustrates how to use [`initalize_ssl_context()`](common/openssl_utility.cpp#L118) to configure the `SSL_CTX` as suggested in both the [server](server/enc/openssl_server.cpp#L147) and [client](client/enc/openssl_client.cpp#L200) modules. 
```

  ##### Recommended configurations for TLS communication

>   - TLS protocol versions

Reviewer: line 99-101 can be removed.

    - TLS 1.2
    - TLS 1.3

>   - TLS 1.3 cipher suites:

Reviewer: ```suggestion
  - TLS 1.3 cipher suites (with the exact order)
```

    - TLS13-AES-256-GCM-SHA384
    - TLS13-AES-128-GCM-SHA256

>   - TLS 1.2 cipher suites:

Reviewer: ```suggestion
  - TLS 1.2 cipher suites (with the exact order)
```

```cmd
.\server\host\tls_server_host .\server\enc\tls_server_enc.signed -port:12341 -server-in-loop
```
### Recommended TLS configurations when using OpenSSL

>   It is strongly recommended that developers configure OpenSSL to restrict the TLS versions, ciphersuites and elliptic curves to be used for TLS connections to enclave:

Reviewer: ```suggestion
  It is strongly recommended that developers configure OpenSSL to restrict the TLS versions, cipher suites, and elliptic curves to be used for TLS connections to enclave:
```

  - TLS protocol versions
    - TLS 1.2
    - TLS 1.3
  - TLS 1.3 cipher suites (in the exact order)
    - TLS13-AES-256-GCM-SHA384
    - TLS13-AES-128-GCM-SHA256
  - TLS 1.2 cipher suites (in the exact order)
    - ECDHE-ECDSA-AES128-GCM-SHA256
    - ECDHE-ECDSA-AES256-GCM-SHA384
    - ECDHE-RSA-"AES128-GCM-SHA256
    - ECDHE-RSA-AES256-GCM-SHA384
    - ECDHE-ECDSA-AES128-SHA256
    - ECDHE-ECDSA-AES256-SHA384
    - ECDHE-RSA-AES128-SHA256
    - ECDHE-RSA-AES256-SHA384
  - Elliptic curves
    - P-521
    - P-384
    - P-256

>   This sample illustrates how to use [`initalize_ssl_context()`](common/openssl_utility.cpp#L118) to configure the `SSL_CTX` as suggested in both the [server](server/enc/openssl_server.cpp#L147) and [client](client/enc/openssl_client.cpp#L200) modules.

Reviewer: ```suggestion
  This sample illustrates how to use [`initalize_ssl_context()`](common/openssl_utility.cpp#L118) to configure the `SSL_CTX` as suggested in both the [server](server/enc/openssl_server.cpp#L147) and the [client](client/enc/openssl_client.cpp#L200) modules.
```
